routing/payment_lifecycle.go

```diff
@@ -26,189 +28,524 @@ func (e errNoRoute) Error() string {
+		}
+
+		var activeShards int
+		var sentAmt lnwire.MilliSatoshi
+		for _, a := range attempts {
+			if a.Settle != nil || a.Failure != nil {
+				continue
 			}
 
-			// Now that the attempt is created and checkpointed to
-			// the DB, we send it.
-			sendErr := p.sendPaymentAttempt(firstHop, htlcAdd)
-			if sendErr != nil {
-				// TODO(joostjager): Distinguish unexpected
-				// internal errors from real send errors.
-				err = p.failAttempt(sendErr)
-				if err != nil {
-					return [32]byte{}, nil, err
-				}
+			activeShards++
+			sentAmt += a.Route.Amt()
+		}
 
-				// We must inspect the error to know whether it
-				// was critical or not, to decide whether we
-				// should continue trying.
-				err := p.handleSendError(sendErr)
-				if err != nil {
-					return [32]byte{}, nil, err
+		// The payment is finished if we have no outstanding shards,
+		// and we recorded either a success or a failure.
+		if activeShards == 0 && (success || failure != nil) {
+			// Find the first successful shard and return the
+			// preimage and route.
+			for _, a := range attempts {
+				if a.Settle != nil {
+					return a.Settle.Preimage, &a.Route, nil
 				}
 
-				// Error was handled successfully, reset the
-				// attempt to indicate we want to make a new
-				// attempt.
-				p.attempt = nil
-				continue
 			}
-		} else {
-			// If this was a resumed attempt, we must regenerate the
-			// circuit. We don't need to check for errors resulting
-			// from an invalid route, because the sphinx packet has
-			// been successfully generated before.
-			_, c, err := generateSphinxPacket(
-				&p.attempt.Route, p.payment.PaymentHash[:],
-				p.attempt.SessionKey,
-			)
-			if err != nil {
-				return [32]byte{}, nil, err
+
+			// Payment failed, mark it as as permanently failed
+			// with the control tower.
+			if failure != nil {
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failure.failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
+
+				// Terminal state, return.
+				return [32]byte{}, nil, failure.err
 			}
-			p.circuit = c
-		}
 
-		// Using the created circuit, initialize the error decrypter so we can
-		// parse+decode any failures incurred by this payment within the
-		// switch.
-		errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
-			OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(p.circuit),
+			return [32]byte{}, nil, fmt.Errorf("No successful " +
+				"attempts nor failure found!")
 		}
 
-		// Now ask the switch to return the result of the payment when
-		// available.
-		resultChan, err := p.router.cfg.Payer.GetPaymentResult(
-			p.attempt.AttemptID, p.payment.PaymentHash, errorDecryptor,
-		)
-		switch {
+		remValue := p.totalAmount - sentAmt
+
+	ShardLoop:
+		// While we still have amount to send, and we didn't aleady
+		// encounter a success or parmanent failure, send another
+		// shard.
+		for remValue > 0 && !(success || failure != nil) {
```

Comment:
After the if above, it seems that `!(success || failure != nil)` is always true here. Those values aren't updated inside this for loop either, so maybe the condition can be removed entirely.

Reply:
We still have the case where we have a terminal condition but still have outstanding shard, in that case we don't want to send more even though we have more `remValue`.... or am I crazy? 🤔

Reply:
I mean, if nothing changes `success` or `failure` inside the loop, what is the point of checking it for every iteration?

Reply:
Ah it is just a condition to enter the loop in the first place. Could make an `if` outside, but did it to avoid indentation

Reply:
Yes, I think extracting is better to prevent the question that came up with me. Also this function is 220 lines, so it is probably worth thinking about splitting anyway.

Reply:
Not applicable anymore.